File: SpriteIlluminatorCocos2dObjC.spritebuilder/Source/MainScene.m
/**
 * SpriteIlluminator Demo for Cocos2D
 *
 * Written by Andreas Loew / CodeAndWeb GmbH
 * 
 * This demo shows different 2d dynamic lighting scenarios with a single sprite.
 * The sprite's normal map was created using SpriteIlluminator
 * https://www.codeandweb.com/spriteilluminator
 *
 */
#import "MainScene.h"

@implementation MainScene
{
    CCTime nextFlicker;
}

- (id)init
{
    if (self = [super init])
    {
        self.userInteractionEnabled = TRUE;
        nextFlicker = 0.1;
    }

    return self;
}

-(void) onEnter
{
    // start with default screen
    [self noLightClicked];
    [super onEnter];
}

-(void)touchBegan:(CCTouch *)touch withEvent:(CCTouchEvent *)event
{
    // touch moved only works with touchBegan present
}

- (void)touchMoved:(CCTouch *)touch withEvent:(CCTouchEvent *)event
{
    CGPoint touchLocation = [touch locationInNode:self];
    
    // calculate angle for the directional light
    CGPoint vector = ccpSub(touchLocation,_characterSprite.position);
    float angle = ccpToAngle(vector) * 180.0f /  M_PI;
    _lightNode.rotation = 90-angle;
    _directionalLightCursor.rotation = 180-angle;

    // set positions for light and emitters
    _pointLightCursor.position = _lightNode.position = _fireEmitter.position = touchLocation;
}

-(void) showFire:(BOOL)show
{
    if(show)
    {
        [_fireEmitter resetSystem];
    }
    else
    {
        [_fireEmitter stopSystem];
    }
    
    [_fireEmitter setVisible: show];
}


-(void) showSnow:(BOOL)show
{
    if(show)
    {
        [_snowEmitter resetSystem];
    }
    else
    {
        [_snowEmitter stopSystem];
    }
    [_snowEmitter setVisible: show];
}


-(void) pointLightClicked
{
    // black background
    _background.color = [CCColor colorWithRed:0.0f green:0.0f blue:0.0f];

    // white point light with gray ambient light
    _lightNode.color = [CCColor colorWithRed:1.0f green:1.0f blue:1.0f];
    _lightNode.ambientColor = [CCColor colorWithRed:0.5f green:0.5f blue:0.5f];
    _lightNode.type = CCLightPoint;
    _lightNode.position =_pointLightCursor.position = ccp(400,260);
    
    _pointLightCursor.visible = TRUE;
    _directionalLightCursor.visible = FALSE;
    _characterWithoutLighting.visible = FALSE;
    _characterSprite.visible = TRUE;
    
    [self showFire:NO];
    [self showSnow:NO];
}

-(void) fireClicked
{
    // black background
    _background.color = [CCColor colorWithRed:0.0f green:0.0f blue:0.0f];

    // orange light
    // dark blue ambient light
    // animated in update:
    _lightNode.color = [CCColor colorWithRed:1.0f green:0.5f blue:0.5f];
    _lightNode.ambientColor = [CCColor colorWithRed:0.4f green:0.4f blue:0.8f];
    _lightNode.type = CCLightPoint;
    _lightNode.position = _fireEmitter.position = ccp(400,60);
    
    _pointLightCursor.visible = FALSE;
    _directionalLightCursor.visible = FALSE;
    _characterWithoutLighting.visible = FALSE;
    _characterSprite.visible = TRUE;

    [self showFire:YES];
    [self showSnow:NO];
}

-(void) summerClicked
{
    // green background
    _background.color = [CCColor colorWithRed:0.6f green:0.8f blue:0.6f];

    // bright light with a touch of warm yello
    _lightNode.color = [CCColor colorWithRed:1.0f green:1.0f blue:0.8f];
    _lightNode.ambientColor = [CCColor colorWithRed:0.8f green:0.8f blue:0.8f];
    _lightNode.type = CCLightDirectional;
    _lightNode.depth = 1.0;
    _lightNode.rotation = 0;
    
    _pointLightCursor.visible = FALSE;
    _directionalLightCursor.visible = TRUE;
    _characterWithoutLighting.visible = FALSE;
    _characterSprite.visible = TRUE;

    [self showFire:NO];
    [self showSnow:NO];
}

-(void) winterClicked
{
    // cold blue background
    _background.color = [CCColor colorWithRed:0.5f green:0.7f blue:1.0f];

    // bright light, with some icy blue
    _lightNode.color = [CCColor colorWithRed:0.7f green:0.9f blue:1.0f];
    _lightNode.ambientColor = [CCColor colorWithRed:0.6f green:0.7f blue:1.0f];
    _lightNode.type = CCLightDirectional;
    _lightNode.depth = 1.0;
    _lightNode.rotation = 0;
    
    _pointLightCursor.visible = FALSE;
    _directionalLightCursor.visible = TRUE;
    _characterWithoutLighting.visible = FALSE;
    _characterSprite.visible = TRUE;
    
    [self showFire:NO];
    [self showSnow:YES];
}

-(void) noLightClicked
{
    // dark blue background, no light set on this sprite
    _background.color = [CCColor colorWithRed:0.1f green:0.1f blue:0.3f];

    _characterWithoutLighting.visible = TRUE;
    _characterSprite.visible = FALSE;
    _pointLightCursor.visible = FALSE;
    _directionalLightCursor.visible = FALSE;

    [self showFire:NO];
    [self showSnow:NO];
}

-(void) update:(CCTime)delta
{
    if(_fireEmitter.visible)
    {
        nextFlicker -= delta;
        if(nextFlicker < 0)
        {
            nextFlicker = 0.1;
            // choose random position for light node
            // close to fire emmitter
            _lightNode.position = ccpAdd(_fireEmitter.position, ccp(rand() % 30 - 15, rand() % 30 - 15));
        }
    }
}

@end

File: SpriteIlluminatorCocos2dSwift.spritebuilder/Source/MainScene.swift
import Foundation

class MainScene: CCNode
{
    var  pointLightCursor:CCSprite!
    var  directionalLightCursor:CCSprite!
    var  lightNode:CCLightNode!
    var  fireEmitter:CCParticleSystem!
    var  snowEmitter:CCParticleSystem!
    var  background:CCNodeColor!
    var  characterSprite:CCSprite!
    var  characterWithoutLighting:CCSprite!
    
    var nextFlicker:CCTime
    
    override init()
    {
        nextFlicker = 0.1
        super.init()
        self.userInteractionEnabled = true
    }
    
    override func onEnter()
    {
        noLightClicked()
        super.onEnter()
    }
    
    override func touchBegan(touch: CCTouch, withEvent event: CCTouchEvent)
    {
        // touch moved only works with touchBegan present
    }
    
    override func touchMoved(touch: CCTouch!, withEvent event: CCTouchEvent!)
    {
        var touchLocation = touch.locationInNode(self)
        
        // calculate angle for the directional light
        var vector = ccpSub(touchLocation, characterSprite.position)
        var angle = Float(Double(ccpToAngle(vector)) * 180.0 / M_PI)
        lightNode.rotation = 90 - angle
        directionalLightCursor.rotation = 180 - angle
        
        // set positions for light and emitters
        pointLightCursor.position = touchLocation
        lightNode.position = touchLocation
        fireEmitter.position = touchLocation
    }
    
    func showFire(show:Bool)
    {
        if (show)
        {
            fireEmitter.resetSystem()
        }
        else
        {
            fireEmitter.stopSystem()
        }
        fireEmitter.visible = show
    }
    
    func showSnow(show:Bool)
    {
        if (show)
        {
            snowEmitter.resetSystem()
        }
        else
        {
            snowEmitter.stopSystem()
        }
        snowEmitter.visible = show
    }
    
    func pointLightClicked()
    {
        // black background
        background.color = CCColor(red: 0.0, green: 0.0, blue: 0.0)
        
        // white point light with gray ambient light
        lightNode.color = CCColor(red: 1.0, green: 1.0, blue: 1.0)
        lightNode.ambientColor = CCColor(red: 0.5, green: 0.5, blue: 0.5)
        lightNode.type = CCLightType.Point
        lightNode.position  = ccp(400, 260)
        pointLightCursor.position = lightNode.position
        
        pointLightCursor.visible = true
        directionalLightCursor.visible = false
        characterWithoutLighting.visible = false
        characterSprite.visible = true
        
        showFire(false)
        showSnow(false)
    }
    
    func fireClicked()
    {
        // black background
        background.color = CCColor(red: 0.0, green: 0.0, blue: 0.0)
        
        // orange light
        // dark blue ambient light
        // animated in update:
        lightNode.color = CCColor(red: 1.0, green: 0.5, blue: 0.5)
        lightNode.ambientColor = CCColor(red: 0.4, green: 0.4, blue: 0.8)
        lightNode.type = CCLightType.Point
        lightNode.position = ccp(400, 60)
        fireEmitter.position = lightNode.position
        
        pointLightCursor.visible = false
        directionalLightCursor.visible = false
        
        characterWithoutLighting.visible = false
        characterSprite.visible = true
        
        showFire(true)
        showSnow(false)
    }
    
    func summerClicked()
    {
        // green background
        background.color = CCColor(red: 0.6, green: 0.8, blue: 0.6)
        
        // bright light with a touch of warm yello
        lightNode.color = CCColor(red: 1.0, green: 1.0, blue: 0.8)
        lightNode.ambientColor = CCColor(red: 0.8, green: 0.8, blue: 0.8)
        lightNode.type = CCLightType.Directional
        lightNode.depth = 1.0
        lightNode.rotation = 0
        
        pointLightCursor.visible = false
        directionalLightCursor.visible = true
        characterWithoutLighting.visible = false
        characterSprite.visible = true
        
        showFire(false)
        showSnow(false)
    }
    
    func winterClicked()
    {
        // cold blue background
        background.color = CCColor(red: 0.5, green: 0.7, blue: 1.0)
        
        // bright light, with some icy blue
        lightNode.color = CCColor(red: 0.7, green: 0.9, blue: 1.0)
        lightNode.ambientColor = CCColor(red: 0.6, green: 0.7, blue: 1.0)
        lightNode.type = CCLightType.Directional
        lightNode.depth = 1.0
        lightNode.rotation = 0;
        
        pointLightCursor.visible = false
        directionalLightCursor.visible = true
        characterWithoutLighting.visible = false
        characterSprite.visible = true
        
        showFire(false)
        showSnow(true)
    }
    
    func noLightClicked()
    {
        // dark blue background, no light set on this sprite
        background.color = CCColor(red: 0.1, green: 0.1, blue: 0.3)
        
        characterWithoutLighting.visible = true
        characterSprite.visible = false
        pointLightCursor.visible = false
        directionalLightCursor.visible = false
        
        showFire(false)
        showSnow(false)
    }
    
    override func update(delta: CCTime)
    {
        if(fireEmitter.visible)
        {
            nextFlicker -= delta
            if(nextFlicker < 0)
            {
                nextFlicker = 0.1
                // choose random position for light node
                // close to fire emmitter
                lightNode.position = ccpAdd(fireEmitter.position, CGPoint(x: rand() % 30 - 15, y:rand() % 30 - 15))
            }
        }
    }
}
